docs/developer-docs/security/rust-canister-development-security-best-practices.md

```diff
@@ -262,8 +262,151 @@ Every inter-canister call is guaranteed to receive a response, either from the c
+The fact that the intent has been journaled and the app knows not to reenter the flow until the result has been recorded means the journal acts as a lock on the critical section containing
+the ledger outcall. The lock will not get stuck assuming the application can always find out what happened to a call. Enough context about the call should be recorded in the intent to ensure
+that this is the case. For the ICP ledger, an ID can be generated and recorded in the journaled intent and the ledger can be called with the ID included in the memo so that the result of the
+call can be queried later.
+
+#### Journaling is Robust to Panics
+
+Continuing the above example, let us consider a panic at any point. 
+1. If there is panic before the async outcall, then the journaled intent will be lost, but no state change occurred internally and no outcalls were made so the app is in a safe state. The next step is to record a new intent.
+1. If there is a panic after the async outcall, then the journaled result (step 3) will be lost. This means the app will need to determine the result and journal it before continuing to step 4. As long as it is possible to determine the result, the app is in a safe state even.
+
+#### Journaling And Audit Events
+
+The journal can be used to augment the audit trail for recent events. However, it is probably too detailed for long term storage. After a while, journal entries could be compressed and incorporated into long term audit events. The process for creating audit events could itself be journaled.
+
+#### Recovery
+
+The journal ensures the application knows that recovery from an error is needed, and aids in making recovery decisions. In order to support the recovery process, the journal should support querying all unresolved tasks of a certain type, and tasks of a certain type that resulted in an error. Given an intent, the journal should also be able to return the result if it exists and indicate if it does not exist.
+
+Extending the ledger example above, the recovery process can look like the following
+
+1. There is a panic and the status of the ledger call is unknown. However, the journal has recorded that a call to transfer with particular parameters and memo has been made, and the deduplication timestamp of the transfer. 
+1. The app calls the ledger to determine whether a transaction with the journaled parameters has succeeded on the ledger. Note that with the ICP ledger, there is a small possibility the transaction is stuck in the message queue, so one needs to wait 24 hours from the deduplication time stamp to ensure the message has either succeeded or failed and is not in transit. Therefore if the ledger does not contain the transaction and it is still within 24 hours of the deduplication timestamp, this step must be repeated.
```

Comment:
The transaction could be also stuck at some point before entering the message queue. In that case, it seems this step is still necessary. If we agree, I will change the text to describe the case where the transaction message is stuck before entering the ledger's message queue.

Reply:
I see, I misunderstood your earlier message as being that execution of messages is the same as ordering in the message queue.

That aside, I'd rephrase the question of interest as: Does the guaranteed ordering apply all the way from the caller message execution to the callee message execution? I.e. does "(order of caller execution of a message `M1[i]` that makes a call `M2[i]`) = (order of callee message execution of `M2[i]`)"? It sounds like is the case, but I would like to confirm, since it is a distinct statement from "order of execution = order of sending assuming both messages are executed".